### ISSUE TITLE
Catch undefined command / function

### ISSUE THREAD

**Issue by functicons** [koalaman/shellcheck#2574]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

Is it possible to generate warning for undefined command / function? The use case is to catch typos. In practice, I'm okay to maintain a dictionary of command / function names.

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

echa "hello world"

```

#### Here's what shellcheck currently says:

`No issues detected!`

#### Here's what I wanted or expected to see:

`Unknown command "echa"`

**Comment by brother** [comment]
maybe a sounds like match to catch common misspellings or such. it's hard to know what binaries are installed in the PATH of the system and so on.

**Comment by christophgil** [comment]
Shellcheck is extremly helpful. Thanks for this!

I would highly vote for -  Indicating misspelled functions/shell commands.

However, searching executable paths  may be time consuming.
Furthermore, it will also yield false positives when the dev-environment is not the target machine.

Therefore I think that searching for misspelled commands/functions should not be the default.
It should be activated by a command line option.

Many Thanks!

**Comment by Commenter25** [comment]
Would also find this very useful, particularly for functions. Perhaps an alternative to searching the PATH: pairing it with a directive to state a list of binaries expected on the system?

I imagine it'd primarily be used file-wide or in `.shellcheckrc`, but it could also be used like an assertion that the script has already verified the binaries exist. Something like this:

```bash
#!/usr/bin/bash
if ! hash jq awk grep; then
    exit 1
fi

# shellcheck binaries=grep,jq,awk
grep -m 1 VERSION /etc/os-release | awk -F "=" '{ print $2 }' | jq '@uri'
```

I made `awk` and `grep` explicit here for demonstration, but realistically, it should automatically include [the standard commands](https://en.wikipedia.org/wiki/List_of_POSIX_commands) you'd expect on a POSIX system, and any builtins for the specified shell.

Outside of that, it can assume anything not specified in a `binaries` directive doesn't exist, unless it's in an explicit if check with `command -v` like above. This resolves the ambiguity conflict between functions and commands, allowing functions to always be properly checked for existence.

Below is a hypothetical wiki page for my proposed rule.

---

## Are you certain foo exists? Use a 'binaries' directive, check with `command -v`, or define a function.

### Problematic code:

```bash
echo "Hello World" | foo  # foo might not exist!
```

### Correct code: (with directive)

```bash
# shellcheck binaries=foo # trust me, foo exists
echo "Hello World" | foo
```

### Correct code: (with error checking)

```bash
if command -v foo; then
    echo "Hello World" | foo
else
    echo "Please install foo"
    exit 1
fi
```

### Correct code: (with function)

```bash
foo() {
    sed 's/World/Bar/'
}

echo "Hello World" | foo
```

### Rationale:

Outside of shell builtins and [standard POSIX commands](https://en.wikipedia.org/wiki/List_of_POSIX_commands), scripts meant for distribution should ensure a command exists before calling it. 

If you intended to call an external binary, you can use a [`binaries` directive](https://github.com/koalaman/shellcheck/wiki/Directive#binaries) to specify that you expect it to be available, or you can add an explicit runtime check using `command -v foo`.

Alternatively, you may have been trying to reference a function that is undefined. In this case, double check that the function is indeed defined, and that the name is not misspelled.

### Exceptions:

If you know the target environment and what binaries are available, such as a script for personal use, you can [ignore](https://www.shellcheck.net/wiki/ignore) this warning. Note that this will also suppress undefined function warnings, so a file-wide `binaries` directive may be preferable.

Note that ShellCheck intentionally does not attempt to figure out runtime or dynamic assignments like with `source "$(date +%F).sh"` or `eval "foo() { bar }"`. See [SC2034](https://www.shellcheck.net/wiki/SC2034) for an extended discussion of why this is the case.

### Related resources:

If you wish to explicitly check for binaries, the optional rule `deprecate-which` and the information on its wiki page may be of interest: https://www.shellcheck.net/wiki/SC2230

* Help by adding links to BashFAQ, StackOverflow, man pages, POSIX, etc!

**Comment by ArrayBolt3** [comment]
> If you intended to call an external binary, you can use a [binaries directive](https://github.com/koalaman/shellcheck/wiki/Directive#binaries) to specify that you expect it to be available, or you can add an explicit runtime check using command -v foo.

Note that under Bash `command -v` doesn't check if the file it finds is actually executable:

```
user@ks-dev:~$ command -v zip
/usr/bin/zip
user@ks-dev:~$ sudo chmod -x /usr/bin/zip
user@ks-dev:~$ command -v zip
/usr/bin/zip
```

This behavior can cause some fun issues if your system has the world-executable bit unset for some root-owned binaries. One has to combine `command -v` with a `test -x`  check to be certain. zsh's `command -v` doesn't suffer from this.
